### PR TITLE
Return allowed `kube_users` and `kube_groups` that can be used to access a certain `kube_cluster` and allow custom request impersonation

### LINKED ISSUE #17382: Add fields users and groups for each kube cluster when querying with `ListResources`
For the `kube` discovery flow, we need a way to receive current kube users and groups (static + dynamic).
By static I mean the users and groups defined in roles like this:

```yaml
kind: role
spec:
  allow:
    kubernetes_groups:
    # dynamic
    - '{{internal.kubernetes_groups}}'
    # static groups
    - kubeGroup1
    - kubeGroup2
    kubernetes_users:
    # dynamic
    - '{{internal.kubernetes_users}}'
    # static users
    - kubeUser1
    - kubeUser2
 ```

I think we can use the same technique used when [querying for list of nodes](https://github.com/gravitational/teleport/blob/9f9461d5f0186800c184345d9102457fc28d4f3a/lib/web/apiserver.go#L2093), where for each node, we [EnumerateServerLogins](https://github.com/gravitational/teleport/blob/9f9461d5f0186800c184345d9102457fc28d4f3a/lib/services/role.go#L950) and add it as a field called [SSHLogins](https://github.com/gravitational/teleport/blob/9f9461d5f0186800c184345d9102457fc28d4f3a/lib/web/ui/server.go#L52). 

Similar references to kube
- [kube api](https://github.com/gravitational/teleport/blob/9f9461d5f0186800c184345d9102457fc28d4f3a/lib/web/servers.go#L30) used to query for kube clusters (that should now also return the users and groups)
- where we should [enumerate](https://github.com/gravitational/teleport/blob/9f9461d5f0186800c184345d9102457fc28d4f3a/lib/web/ui/server.go#L122) for users and groups

### PR DESCRIPTION
**Pull request by tigrato**
This PR returns a list of allowed principals that a user is authorized to impersonate for each Kubernetes cluster.

When a Teleport user has multiple Kubernetes users defined for the same Cluster we must allow him to choose one otherwise the request will return an error. Similar to `kubectl get pods --as {user} --as-group {kubeGroup}`.
If the user selects invalid users or groups, it will return invalid Kubernetes principals.

Closes #17382

### CODE REVIEW

File: lib/services/role.go [hunk 1]
```diff
@@ -948,6 +948,42 @@ func (set RoleSet) EnumerateDatabaseUsers(database types.Database, extraUsers ..
 	return result
 }
 
+// EnumerateKubernetesUsersAndGroups works on a given role set to return a minimal list of
+// configured kubernetes_users and kubernetes_groups for a given kubernetes cluster.
+// The wildcard selector is ignored, since it is not allowed for kubernetes principals.
+func (set RoleSet) EnumerateKubernetesUsersAndGroups(kube types.KubeCluster) (EnumerationResult, EnumerationResult) {
```
Comment (r0mant) [inline lib/services/role.go:954]: This seems very similar to what `CheckKubeGroupsAndUsers` does, which is what forwarder uses to determine which users/groups to pass. I understand the difference is that this operates on a particular cluster, but can't `CheckKubeGroupsAndUsers` be reused somehow? E.g. can't we pass it a matcher that would only match a particular cluster?

File: lib/web/ui/server.go [hunk 2]
```diff
@@ -117,10 +114,15 @@ type KubeCluster struct {
 	Name string `json:"name"`
 	// Labels is a map of static and dynamic labels associated with an kube cluster.
 	Labels []Label `json:"labels"`
+	// TODO: fix comments
```
Comment (r0mant) [inline lib/web/ui/server.go:117]:
```suggestion
```

File: lib/web/ui/server.go [hunk 3]
```diff
@@ -143,15 +145,28 @@ func MakeKubeClusters(clusters []types.KubeCluster) []KubeCluster {
...
+			KubeGroups: nonNilSlice(kubeGroups.Allowed()),
 		})
 	}
 
 	return uiKubeClusters
 }
 
+// nonNilSlice checks if the given slice is nil. If it's the case returns an empty
+// slice, otherwise returns the given slice.
+func nonNilSlice(s1 []string) []string {
```
Comment (r0mant) [inline lib/web/ui/server.go:163]: Why is this needed?

File: lib/web/ui/server.go [hunk 4]
```diff
@@ -117,10 +114,15 @@ type KubeCluster struct {
 	Name string `json:"name"`
 	// Labels is a map of static and dynamic labels associated with an kube cluster.
 	Labels []Label `json:"labels"`
+	// TODO: fix comments
+	// KubeUsers are the list of allowed Kubernetes RBAC users that the user can impersonate.
+	KubeUsers []string `json:"kubernetes_users"`
```
Comment (r0mant) [inline lib/web/ui/server.go:119]: Add `omitempty` json tag here and below?

File: lib/services/role.go [hunk 5]
```diff
@@ -948,6 +948,42 @@ func (set RoleSet) EnumerateDatabaseUsers(database types.Database, extraUsers ..
...
+		kubeGroups = append(kubeGroups, role.GetKubeGroups(types.Deny)...)
+	}
+
+	kubeUsers = apiutils.Deduplicate(kubeUsers)
+	kubeGroups = apiutils.Deduplicate(kubeGroups)
+
+	// check each individual user against the cluster.
+	for _, user := range kubeUsers {
+		err := set.checkAccess(kube, AccessMFAParams{Verified: true}, NewKubeUserMatcher(user))
+		kubeUsersResult.allowedDeniedMap[user] = err == nil
```
Comment (camscale) [inline lib/services/role.go:975]: This ends up hiding errors that are not AccessDenied errors, and can show users/groups that are listed as allowed as denied due to some possibly-transient error when checking access. I can see this is also how it is done in `EnumerateServerLogins`, but perhaps that is also not correct? Should this enumerate function also return an error for the cases where we failed to check access, rather than reporting that as denied?

**Review by camscale — APPROVED** [review]

File: lib/web/ui/server.go (hunk 3 above)
Comment (tigrato) [reply lib/web/ui/server.go:163]: I followed the logic in SSH logins where non-nil arrays are returned. The same applies to `omitempty` where our API does not have any  `omitempty` for teleport resources such as `Database`, `Server`, `Kubes`, and `DesktopService`.

@kimlisa do you have any intel for this?

File: lib/services/role.go (hunk 1 above)
Comment (tigrato) [reply lib/services/role.go:954]: You are correct. I didn't realize the presence of `CheckKubeGroupsAndUsers`. I changed the function and it now uses `CheckKubeGroupsAndUsers` internally with a label matcher.

File: lib/web/ui/server.go (hunk 3 above)
Comment (kimlisa) [reply lib/web/ui/server.go:162]: i don't know the origins unfortunately, and in case of webapps, it's not required. Maybe the person who created it knew that nil slices are encoded as `null` versus an empty array. But a non-nil slice is encoded as empty array? But the frontend does not expect the backend to explicitly set it to an empty array for us, since we can't depend/assume it's never null/undefined.

File: lib/client/conntest/kube.go [hunk 6]
```diff
@@ -250,6 +256,42 @@ func (s KubeConnectionTester) handleErrFromKube(ctx context.Context, clusterName
 			connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
 			return connDiag, trace.Wrap(err)
 		}
+		// WARNING: Check compatibility between this error message in the current version of
```
Comment (r0mant) [inline lib/client/conntest/kube.go:259]: Who is this warning for and what is the call to action here? It's not very clear.

If this depends on some error message strings to be maintained in sync between different versions, can they be moved into constants? Otherwise it's not super clear what should be checked and when.

File: lib/services/role.go [hunk 7]
```diff
@@ -948,6 +948,16 @@ func (set RoleSet) EnumerateDatabaseUsers(database types.Database, extraUsers ..
 	return result
 }
 
+// GetAllowedKubeUsersAndGroupsForCluster works on a given role set to return a minimal list of
+// configured kubernetes_users and kubernetes_groups for a given kubernetes cluster.
+// The wildcard selector is ignored, since it is not allowed for kubernetes principals.
+func (set RoleSet) GetAllowedKubeUsersAndGroupsForCluster(kube types.KubeCluster) (kubeUsers []string, kubeGroups []string) {
+	matcher := NewKubernetesClusterLabelMatcher(kube.GetAllLabels())
+	// ignore error since we are only interested in Allowed  groups
+	kubeGroups, kubeUsers, _ = set.CheckKubeGroupsAndUsers(5*time.Minute, true /* force ttl override*/, matcher)
```
Comment (r0mant) [inline lib/services/role.go:957]: I don't think we want to hardcode the TTL here. It should be a parameter to `GetAllowedKubeUsersAndGroupsForCluster` and should be populated based on user identity's expiration.

**Review by r0mant — COMMENTED** [review]
Couple more comments.

File: lib/services/role.go (hunk 7 above)
Comment (tigrato) [reply lib/services/role.go:957]: My thought was: since users have no control over the UI session duration, it is not possible for them to change the value to be able to receive `kubernetes_users' and `kubernetes_groups' from roles whose MaxSessionTTL is less than the session expiration.  So I wanted to not validate the TTL (override=true) so that they can receive all the groups they can eventually use if their session is less than the MaxSessionTTL value for that role. Otherwise, we could just return values that might not include some groups, which would be strange to the user since they were configured.

**Review by r0mant — APPROVED** [review]
